### PR TITLE
KAFKA-15647: Fix the different behavior in error handling between the old and new group coordinator

### PR DESCRIPTION
**Pull request by dongnuo123**
In `KafkaApis.scala`, we build the API response differently if exceptions are thrown during the API execution. Since the new group coordinator only populates the response with error code instead of throwing an exception when an error occurs, there may be different behavior between the existing group coordinator and the new one.

This pr
- Fixes the response building in `KafkaApis.scala` for the two APIs affected by such difference -- OffsetFetch and OffsetDelete.
- In `GroupCoordinatorService.java`, returns a response with error code instead of a failed future when the coordinator is not active.

### Jira
https://issues.apache.org/jira/browse/KAFKA-15647

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 1]
```diff
@@ -3267,6 +3267,35 @@ class KafkaApisTest {
     assertEquals(Errors.GROUP_ID_NOT_FOUND, Errors.forCode(response.data.errorCode))
   }
 
+  @Test
+  def testOffsetDeleteWithInvalidGroupWithTopLevelError(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:3271]: I think that we don't fully reproduce the bug with this test because `if (data.topics().isEmpty()) {` would be true in this case as well. Would it be possible to have a test which really have partial results awaiting in the builder?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 2]
```diff
@@ -4341,6 +4371,15 @@ class KafkaApisTest {
         false
       )).thenReturn(group3Future)
 
+      val group4Future = new CompletableFuture[OffsetFetchResponseData.OffsetFetchResponseGroup]()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:4381]: I also feel like that we don't fully repro the issue in this one. We also need to test for both the all topics (null) and specific topics (non null) cases. Do you agree?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java [hunk 3]
```diff
@@ -310,16 +313,22 @@ public CompletableFuture<JoinGroupResponseData> joinGroup(
         JoinGroupRequestData request,
         BufferSupplier bufferSupplier
     ) {
+        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+
         if (!isActive.get()) {
-            return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
-        }
+            responseFuture.complete(new JoinGroupResponseData()
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java]: nit: Could we also use `return CompletableFuture.completedFuture`)?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java [hunk 4]
```diff
@@ -310,16 +313,22 @@ public CompletableFuture<JoinGroupResponseData> joinGroup(
...
+                .setMemberId(request.memberId())
+                .setErrorCode(Errors.COORDINATOR_NOT_AVAILABLE.code())
+            );
 
-        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+            return responseFuture;
+        }
 
         if (!isGroupIdNotEmpty(request.groupId())) {
             responseFuture.complete(new JoinGroupResponseData()
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java]: We could also use `return CompletableFuture.completedFuture` here.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorServiceTest.java [hunk 5]
```diff
@@ -139,7 +138,7 @@ public void testStartupShutdown() throws Exception {
     }
 
     @Test
-    public void testConsumerGroupHeartbeatWhenNotStarted() {
+    public void testConsumerGroupHeartbeatWhenNotStarted() throws ExecutionException, InterruptedException {
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorServiceTest.java:141]: Is it worth adding a similar test for each API? That would ensure that we keep the contract.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java [hunk 6]
```diff
@@ -268,7 +269,9 @@ public CompletableFuture<ConsumerGroupHeartbeatResponseData> consumerGroupHeartb
         ConsumerGroupHeartbeatRequestData request
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:269]: Should we also update the javadoc of all those methods to specify the contract a little more? I known that the old coordinator won't fully comply with this but as it will go away, it should not be a big deal.

**Review by dajac — COMMENTED** [review]
@dongnuo123 Thanks for the PR. I left a few small comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java (hunk 6 above)
Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java:269]: Not sure if I understand, should we specify in group coordinator interface or in group coordinator service? I made some change in the group coordinator interface.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 7]
```diff
@@ -4385,15 +4451,39 @@ class KafkaApisTest {
...
+          new OffsetFetchResponseData.OffsetFetchResponseTopics()
+            .setName("foo")
+            .setPartitions(List(
+              new OffsetFetchResponseData.OffsetFetchResponsePartitions()
+                .setPartitionIndex(1)
+                .setCommittedOffset(-1)
+            ).asJava)
+        ).asJava)
+
+      val expectedGroups = List(group1Response, group2Response, group3Response, group4Response, group5Response)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Is this change really necessary?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 8]
```diff
@@ -4385,15 +4451,39 @@ class KafkaApisTest {
         .setGroupId("group-3")
         .setErrorCode(Errors.INVALID_GROUP_ID.code)
 
-      val expectedOffsetFetchResponse = new OffsetFetchResponseData()
-        .setGroups(List(group1Response, group2Response, group3Response).asJava)
+      val group4Response = new OffsetFetchResponseData.OffsetFetchResponseGroup()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:4434]: Actually, how is this different from `group3Response`?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 9]
```diff
@@ -4385,15 +4451,39 @@ class KafkaApisTest {
...
+        .setGroupId("group-4")
+        .setErrorCode(Errors.INVALID_GROUP_ID.code)
+
+      val group5Response = new OffsetFetchResponseData.OffsetFetchResponseGroup()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: I don't fully understand this test case either. My understanding is that we could have a weird response only in one case: the coordinator returns a response with an error and the kafka apis adds unauthorized topics to it. Did I get it right? If so, it may be better to actually add a new (and separate) test case for this one.

**Review by dajac — COMMENTED** [review]
@dongnuo123 Thanks for the update. I left a few more comments.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 8 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:4434]: group3Response and group4Response are the same. I wanted to make sure when
 ```
group3Future.completeExceptionally(Errors.INVALID_GROUP_ID.exception)
group4Future.complete(group4Response)
```
handleOffsetFetch gives the same response.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 9 above)
Comment (dongnuo123) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Yeah, you're right. Let me write it to a separate method. We probably don't need this group 5

**Comment by dajac** [comment]
@dongnuo123 One of the build is still red. You may have to merge trunk to get the latest changes.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 10]
```diff
@@ -1461,6 +1461,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         new OffsetFetchResponseData.OffsetFetchResponseGroup()
           .setGroupId(offsetFetchRequest.groupId)
           .setErrorCode(Errors.forException(exception).code)
+      } else if (offsetFetchResponse.errorCode() != Errors.NONE.code()) {
```
Comment (dajac) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: We could remove the `()` after `code`.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 11]
```diff
@@ -1500,6 +1502,8 @@ class KafkaApis(val requestChannel: RequestChannel,
...
```
Comment (dajac) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: ditto.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 12]
```diff
@@ -4701,6 +4751,89 @@ class KafkaApisTest {
...
+        .setTopics(List(new OffsetFetchRequestData.OffsetFetchRequestTopics()
+          .setName("bar")
+          .setPartitionIndexes(List[Integer](0).asJava)).asJava),
+      false
+    )).thenReturn(group1Future)
+
+    createKafkaApis(authorizer = Some(authorizer)).handle(requestChannelRequest, RequestLocal.NoCaching)
+
+    // group-2 mocks using the new group coordinator.
+    // When the coordinator is not active, a response with error code is returned.
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Should we add a note about `foo` here? The whole point of this test is to ensure that the failed topics are not present in the response when there is a top level error.

**Review by dajac — COMMENTED** [review]
@dongnuo123 Thanks for the update. I left a few nits.

**Review by dajac — APPROVED** [review]
LGTM, thanks. I will merge it when the build completes.
